Public/Src/Cache/ContentStore/Library/Stores/ContentHashAddressableSnapshot.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using BuildXL.Cache.ContentStore.Hashing;
+
+namespace BuildXL.Cache.ContentStore.Stores
+{
+    /// <summary>
+    /// This class is specifically for usage within <see cref="ContentHashAddressableSnapshot{T}"/>. It is a reference type on purpose, in order to avoid issues
+    /// with the maximum object size limit.
+    /// </summary>
+    /// <typeparam name="T">Type tagged with a hash</typeparam>
+    public class PayloadFromDisk<T> : IComparable<PayloadFromDisk<T>>
+    {
+        /// <summary>
+        /// Hash for the <see cref="Payload"/>
+        /// </summary>
+        public readonly ContentHash Hash;
+
+        /// <summary>
+        /// Information for which <see cref="Hash"/> applies
+        /// </summary>
+        public readonly T Payload;
+
+        /// <nodoc />
+        public PayloadFromDisk(ContentHash hash, T payload)
+        {
+            Hash = hash;
+            Payload = payload;
+        }
+
+        /// <nodoc />
+        public int CompareTo(PayloadFromDisk<T> other)
+        {
+            return Hash.CompareTo(other.Hash);
+        }
+    }
+
+    /// <summary>
+    /// This class represents an immutable snapshot of a collection of items at an undetermined point in time, used for enumerations of the file
+    /// system. It is made specifically to avoid issues with enumerations of large amounts of files which go over the maximum object size
+    /// restriction.
+    /// </summary>
+    /// <typeparam name="T">Type held inside the snapshot</typeparam>
+    public class ContentHashAddressableSnapshot<T> : IEnumerable<PayloadFromDisk<T>>
+    {
+        private List<PayloadFromDisk<T>>[] _snapshot;
+
+        /// <nodoc />
+        public readonly long Count = 0;
+
+        /// <nodoc />
+        public ContentHashAddressableSnapshot()
+        {
+            InitializeSnapshot();
+        }
+
+        /// <nodoc />
+        public ContentHashAddressableSnapshot(IEnumerable<PayloadFromDisk<T>> snapshot)
+        {
+            InitializeSnapshot();
+
+            foreach (var payload in snapshot)
+            {
+                var identifier = payload.Hash[0];
+                _snapshot[identifier].Add(payload);
+                Count++;
+            }
+        }
+
+        /// <nodoc />
+        private void InitializeSnapshot()
+        {
+            _snapshot = new List<PayloadFromDisk<T>>[256];
+            for (var i = 0; i < _snapshot.Length; i++)
+            {
+                _snapshot[i] = new List<PayloadFromDisk<T>>();
+            }
+        }
+
+        /// <nodoc />
+        public List<PayloadFromDisk<T>> ListOrderedByHash()
+        {
+            var result = this.ToList();
+            result.Sort();

```

Comment:
This is fine, but another approach may be to keep the entries sorted from the beginning. What do you think about this approach?

Reply:
That approach would incur unnecessary work in one of the use-cases as it stands today. I think it'd be better to keep it like this, and change it in the future if the majority of the workload ends up being this function.

---
In reply to: [285317930](https://github.com/microsoft/BuildXL/pull/329#discussion_r285317930) [](ancestors = 285317930)

Reply:
I see. An alternative would be still to order by each "bucket" individually and yield return the result.

I think in this case the O complexity is still the same, but actual time spend should be smaller.

And as a plus side we can sort all the "buckets" in parallel.

---
In reply to: [285320486](https://github.com/microsoft/BuildXL/pull/329#discussion_r285320486) [](ancestors = 285320486,285317930)

Reply:
These ideas are valid but are getting toward the side of adding WAY too much code. I would like to get the basic structure in place, then develop further if it works and that is justified. Please just sort each individually and append to list. (If you want to factor into a method that yields a sorted IEnumerable and a second method that uses it to produce a sorted list via appending, fine.) More than that is too much.

---
In reply to: [285690964](https://github.com/microsoft/BuildXL/pull/329#discussion_r285690964) [](ancestors = 285690964,285320486,285317930)